packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -823,8 +783,71 @@ export class GarbageCollector implements IGarbageCollector {
         if (!this.tombstoneMode || baseSnapshotData?.tombstones === undefined) {
             return;
         }
-        this.tombstones = baseSnapshotData.tombstones;
-        this.runtime.updateUnusedRoutes(this.tombstones, true /* tombstone */);
+        this.tombstones = Array.from(baseSnapshotData.tombstones);
+        this.runtime.updateTombstonedRoutes(this.tombstones);
+    }
+
+    /**
+     * Update state from the given snapshot data. This is done during load and during refreshing state from a snapshot.
+     */
+    private updateStateFromSnapshotData(snapshotData: IGCSnapshotData) {
+        const currentReferenceTimestampMs = this.runtime.getCurrentReferenceTimestampMs();
+        /**
+         * If there is no current reference timestamp, skip the update. We need the current timestamp to track
+         * how long objects have been unreferenced and if they can be deleted.
+         *
+         * Note that the only scenario where there is no reference timestamp is when no ops have ever been processed
+         * for this container and it is in read mode. In this scenario, there is no point in running GC anyway
+         * because references in the container do not change without any ops, i.e., there is nothing to collect.
+         */
+        if (currentReferenceTimestampMs === undefined) {
+            // Log an event so we can evaluate how often we run into this scenario.
+            this.mc.logger.sendErrorEvent({
+                eventName: "GCStateUpdatedWithoutTimestamp",
+                gcConfigs: JSON.stringify(this.configs),
+            });
+            return;
+        }
+
+        // Clear all existing unreferenced state tracking.
+        this.unreferencedNodesState.forEach((nodeStateTracker: UnreferencedStateTracker) => {
```

Comment:
Nit: `for` loop for consistency? Doesn't really matter to me, feel free to resolve